aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SyncProcessor.java

```diff
@@ -198,13 +213,26 @@ private SyncTime filterOutOldSyncTimes(SyncTime lastSyncTime) throws DataStoreEx
                     "Empty response from AppSync.", "Report to AWS team."
                 ));
             } else {
-                final Set<ModelWithMetadata<T>> emittedValue = new HashSet<>();
-                for (ModelWithMetadata<T> modelWithMetadata : resultFromEndpoint.getData()) {
+                for (ModelWithMetadata<T> modelWithMetadata : resultFromEndpoint.getData().getItems()) {
                     emittedValue.add(modelWithMetadata);
                 }
-                singleEmitter.onSuccess(emittedValue);
+                if (resultFromEndpoint.getData().hasNextResult() && emittedValue.size() < syncMaxRecords()) {
+                    syncPage(resultFromEndpoint.getData().getRequestForNextResult(), singleEmitter, emittedValue);
+                } else {
+                    singleEmitter.onSuccess(emittedValue);
+                }
             }
-        };
+        }, singleEmitter::onError);
+        singleEmitter.setDisposable(AmplifyDisposables.fromCancelable(cancelable));
+    }
+
+    private Integer syncMaxRecords() {
+        try {
+            return dataStoreConfigurationProvider.getConfiguration().getSyncMaxRecords();
+        } catch (DataStoreException exception) {
+            LOG.warn("Failed to retrieve datastore configuration, using default syncMaxRecords value.", exception);
+            return DEFAULT_SYNC_MAX_RECORDS;
+        }
```

Comment:
Is this a realistic scenario?

If so, can we handle this in the code which build the configuration object, instead? S.t. the configuration always has some sane value, by the time we get here to access it?

Reply:
`DataStoreConfigurationProvider::getConfiguration()` does throw a `DataStoreException`, if errors are found building the configuration, but I don't think it is actually thrown in practice.  If the configuration.json is not found, then it uses falls back to defaults.  

I've updated this so that if a `DataStoreException` is actually thrown, it will cause the whole sync to fail, instead of falling back to some default here.  I think this is reasonable, since, as I said, it shouldn't ever be thrown in practice.